src/file_lines.rs

```diff
@@ -0,0 +1,221 @@
+fn normalize_ranges(map: &mut MultiMap<String, Range>) {
+    for (_, ranges) in map.iter_all_mut() {
+        ranges.sort_by_key(|x| x.lo);
+        let merged = ranges.drain(..).coalesce(|x, y| x.merge(y).ok_or((x, y))).collect();
+        *ranges = merged;
+    }
+}
+
+impl FileLines {
+    /// Creates a `FileLines` that contains all lines in all files.
+    pub fn all() -> FileLines {
+        FileLines(None)
+    }
+
+    /// Creates a `FileLines` from a `MultiMap`, ensuring that the invariants hold.
+    fn from_multimap(map: MultiMap<String, Range>) -> FileLines {
+        let mut map = map;
+        normalize_ranges(&mut map);
+        FileLines(Some(map))
+    }
+
+    /// Returns an iterator over the files contained in `self`.
+    pub fn files(&self) -> Files {
+        Files(self.0.as_ref().map(MultiMap::keys))
+    }
+
+    /// Returns true if `range` is fully contained in `self`.
+    pub fn contains(&self, range: &LineRange) -> bool {
+        let map = match self.0 {
+            // `None` means "all lines in all files".
+            None => return true,
+            Some(ref map) => map,
+        };
+
+        match map.get_vec(range.file_name()) {
+            None => false,
+            Some(ranges) => ranges.iter().any(|r| r.contains(Range::from(range))),
+        }
+    }
+
+    /// Returns true if any lines in `range` are in `self`.
+    pub fn intersects(&self, range: &LineRange) -> bool {
+        let map = match self.0 {
+            // `None` means "all lines in all files".
+            None => return true,
+            Some(ref map) => map,
+        };
+
+        match map.get_vec(range.file_name()) {
+            None => false,
+            Some(ranges) => ranges.iter().any(|r| r.intersects(Range::from(range))),
+        }
+    }
+}
+
+/// FileLines files iterator.
+pub struct Files<'a>(Option<::std::collections::hash_map::Keys<'a, String, Vec<Range>>>);
+
+impl<'a> iter::Iterator for Files<'a> {
+    type Item = &'a String;
+
+    fn next(&mut self) -> Option<&'a String> {
+        self.0.as_mut().and_then(Iterator::next)
+    }
+}
+
+// This impl is needed for `Config::override_value` to work for use in tests.
+impl str::FromStr for FileLines {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<FileLines, String> {
+        let v: Vec<JsonSpan> = try!(json::decode(s).map_err(|e| e.to_string()));
+        let m = v.into_iter().map(JsonSpan::into_tuple).collect();
+        Ok(FileLines::from_multimap(m))
+    }
+}
+
+// For JSON decoding.
+#[derive(Clone, Debug, RustcDecodable)]
+struct JsonSpan {
+    file: String,
+    range: (usize, usize),
+}
+
+impl JsonSpan {
+    // To allow `collect()`ing into a `MultiMap`.
+    fn into_tuple(self) -> (String, Range) {
+        let (lo, hi) = self.range;
+        (self.file, Range::new(lo, hi))
+    }
+}
+
+// This impl is needed for inclusion in the `Config` struct. We don't have a toml representation
+// for `FileLines`, so it will just panic instead.
```

Comment:
:+1: Done. I also changed it to have a meaningful panic message. The ideal would be to return an error, but there's now way to make a `D::Error` generically for any `D: Decoder`.

(I will find a way to get something like #795, #801, or something else past you yet!)